src/main/kotlin/no/item/xp/plugin/parser/Parser.kt

```diff
@@ -1,5 +1,60 @@
 package no.item.xp.plugin.parser
 
-import org.gradle.internal.impldep.com.fasterxml.jackson.databind.JsonNode
+import no.item.xp.plugin.models.GeneratedField
+import no.item.xp.plugin.util.NodeListSequence
+import org.w3c.dom.*
 
-interface Parser<T : JsonNode?> : Function<T>
+open class Parser {
+  private var textLineParser: TextLineParser = TextLineParser()
+  companion object Factory {
+    fun create(): Parser = Parser()
+  }
+
+  private fun parseXmlNode(node: Node): Sequence<GeneratedField> {
+    val inputNodes: Sequence<Node> = getInputNodes(node)
+    inputNodes.map {
+      when (getType(it)) {
+        "textline" -> textLineParser.parseTextLine(it)
+        "textarea" -> textLineParser.parseTextLine(it)
+        "datetime" -> textLineParser.parseTextLine(it)
+        "date" -> textLineParser.parseTextLine(it)
+        "" -> textLineParser.parseTextLine(it)
+        else -> println("unable to recognize type")
+      }
+    }
+    return emptySequence()
+  }
+
+  private fun getInputNodes(node: Node): Sequence<Node> {
+    return NodeListSequence(node.childNodes)
+  }
+  private fun getType(node: Node): String? {
+    if (node.nodeName == "type") {
+      return node.nodeValue.toString().toLowerCase()
+    }
+    return null
+  }
+  fun parse(doc: Document?): ArrayList<GeneratedField> {
```

Comment:
Can we rewrite all of these as pure functions (e.g don't have mutable state that is updated)? Put the main function (`parse`) on top of the file (so that it is easy to see where the functionality starts).

Then order the other functions in the order that they are called.